vulcanai2/plotters/utils.py

```diff
@@ -22,66 +23,81 @@ def __init__(self, network):
         else:
             first_layer = self.network.network[0]._kernel
-
         self.hooks.append(first_layer.register_backward_hook(hook_function))
 
-    def crop_negative_gradients(self):
+    def _crop_negative_gradients(self):
         """
-            Updates relu activation functions so that it only returns positive gradients
+            Updates relu/selu activation functions so that it
+            only returns positive gradients
         """
         def activation_hook_function(module, grad_in, grad_out):
             """
             If there is a negative gradient, changes it to zero
             """
             if isinstance(module, ReLU) or isinstance(module, SELU):
                 return (torch.clamp(grad_in[0], min=0.0),)
-        # Loop through layers, hook up activations with activation_hook_function
+        # Since all layer activations in a Net object point to the same
+        # function we only need to hook one of them with
+        # activation_hook_function
         if '_input_network' in self.network._modules:
-            for layer in self.network._input_network.network:
-                self.hooks.append(layer._activation.
-                    register_backward_hook(activation_hook_function))
-        for layer in self.network.network:
-            self.hooks.append(layer._activation.
+            self.hooks.append(self.network._input_network.network[0]._activation.
                 register_backward_hook(activation_hook_function))
+        self.hooks.append(self.network.network[0]._activation.
+            register_backward_hook(activation_hook_function))
+
 
-    def remove_hooks(self):
+    def _remove_hooks(self):
         """
         Remove all previously placed activation hooks from model.
         :return: None
         """
         for h in self.hooks:
             h.remove()
 
-    def generate_gradients(self, input_x, target_class_list):
+    def generate_gradients(self, input_data, targets):
         """
         Computes guided backprop gradients and returns top layer gradients.
 
-        :param input_x: 1D for DenseNet, 4D (for 2D images) or 5D (for 3D images) Tensor.
-        :param target_class_list: 1D list of class truths
-        :return: Gradient array with same shape as input images
+        :param input_data: 1D for DenseNet, 4D (for 2D images) or 5D (for 3D images) Tensor.
+        :param targets: 1D list of class truths of type torch.LongTensor
+        :return: Gradient numpy array with same shape as input images
         """
+        self.network.eval()
+        # To properly pass the gradients
+        if not isinstance(input_data, torch.Tensor):
+            input_data = torch.tensor(input_data, requires_grad=True)
+        else:
+            if not input_data.requires_grad:
+                input_data.requires_grad = True
         # Forward pass
-        network_output = self.network(input_x)
+        network_output = self.network(input_data)
```

Comment:
Change this to: `network_output = self.network.cpu()(input_data)`

Reply:
will do